ats/pkb/pkb.py

```diff
@@ -45,17 +49,82 @@ def process_relations(node: nodes.ASTNode):
+                nodes_stack.append(child)
                 process_relations(child)
 
+                if isinstance(nodes_stack[-1], nodes.VariableNode):
+                    if nodes_stack[-1].parent.condition == nodes_stack[-1]:
+                        uses[nodes_stack[-1].parent.__stmt_id] = [nodes_stack[-1].name]
+                    else:
+                        variable_list.append(nodes_stack[-1].name)
+
+                if isinstance(nodes_stack[-1].parent, nodes.StmtAssignNode):
+                    if nodes_stack[-1].parent.variable == nodes_stack[-1]:
+                        modifies[nodes_stack[-1].parent.__stmt_id] = [
+                            nodes_stack[-1].name
+                        ]
+                    else:
+                        uses[nodes_stack[-1].parent.__stmt_id] = variable_list
+                    variable_list = []
+
+                if isinstance(nodes_stack[-1], nodes.StmtCallNode):
+                    modifies[nodes_stack[-1].__stmt_id] = [nodes_stack[-1]]
+                    uses[nodes_stack[-1].__stmt_id] = [nodes_stack[-1]]
+
+                if isinstance(nodes_stack[-1].parent, nodes.StmtLstNode):
+                    parent_node = nodes_stack[-1].parent
+                    if isinstance(parent_node.parent, nodes.StmtNode):
+                        modifies[parent_node.parent.__stmt_id].extend(
+                            modifies[nodes_stack[-1].__stmt_id]
+                        )
+                        uses[parent_node.parent.__stmt_id].extend(
+                            uses[nodes_stack[-1].__stmt_id]
+                        )
+                    if isinstance(parent_node.parent, nodes.ProcedureNode):
+                        modifies[parent_node.parent.name].extend(
+                            modifies[nodes_stack[-1].__stmt_id]
+                        )
+                        uses[parent_node.parent.name].extend(
+                            uses[nodes_stack[-1].__stmt_id]
+                        )
+
+                nodes_stack.pop()
+
+        variable_list = []
+        nodes_stack.append(tree)
         process_relations(tree)
 
+    nodes_stack = []
     find_all_statements()
     process_all_relations()
 
+    for key, value in modifies.items():
+        variable_list = []
+        for element in value:
+            if isinstance(element, nodes.StmtCallNode):
+                if key != element.name:
+                    variable_list.extend(modifies[element.name])
+            else:
+                variable_list.append(element)
+        modifies[key] = list(set(variable_list))
+
+    for key, value in uses.items():
+        variable_list = []
+        for element in value:
+            if isinstance(element, nodes.StmtCallNode):
+                if key != element.name:
+                    variable_list.extend(uses[element.name])
+            else:
+                variable_list.append(element)
+        uses[key] = list(set(variable_list))
```

Comment:
?

Reply:
Tutaj program przegląda bibliotekę i szuka w bibliotece elementów które są typu procedure node i wtedy zamienia ten element na zmienne z których korzysta ta procedura. Nie wiem czy dobrze zrobiłem ale jak natrafię na statement call to dodaje do biblioteki procedure node który się tak nazywa. Wcześniej próbowałem zrobić, że jak natrafiam na statement call to dodawałem do dict[stmtcall id] = dict[nazwa procedury] i wtedy przy dodawaniu dict[nazwa procedury] wartości automatycznie dodawało się w innych miejscach w bibliotece tam gdzie trzeba było tylko to nie działało gdy procedura wywoływała samą siebie.

Reply:
Daloby sie zrezygnowac z tych forow it odrazu zapisac do `uses` i `modifies` odpowiednie elementy bezposrednio w DFSie?

Reply:
Właśnie chyba nie bo idąc po DFSie najpierw znajdujemy miejsca w bibliotece w których trzeba dodać wartości z kluczem z nazwą procedury, a dopiero potem przypisujemy wartości do klucza z nazwą procedury. Można próbować z jakimiś wskaźnikami, ale jak wcześniej próbowałem to nie działa jak procedura wywołuje samą siebie a i tak trzeba będzie przelecieć po słowniku żeby go znormalizować bo wartość tych kluczy których to będzie dotyczyć będzie wyglądać tak: ['a', b', 'c', ['c', 'e', 'a']] więc trzeba to spłaszczyć do jednej listy i usunąć powtórzenia.

Reply:
> nie działa jak procedura wywołuje samą siebie 

ale ten jezyk nie wspiera rekurencji.

> będzie wyglądać tak: ['a', b', 'c', ['c', 'e', 'a']]

to nie mozna odrazu budowac plaskiej listy? Albo seta zamiast listy? Wtedy nie bedzie powtorzen

Reply:
Nie można od razu płaskiej listy bo mając np taki kod:
```
procedure test1 {
   b = c;
   while a {
      call test2;
   }
}
procedure test2 {
   ...
}
```
DFS najpierw przechodzi przez procedurę test1 zakładając że robimy słownik uses będzie on wyglądał po kolei tak:
uses[1] = ['c']

uses[1] = ['c']
uses[2] = ['a']

uses[1] = ['c']
uses[2] = ['a']
uses[3] = [[]] <- to jest uses[3] = uses['test2']

uses[1] = ['c']
uses[2] = ['a', []] <- to jest ta sama tablica co w uses[3]
uses[3] = [[]]

uses[1] = ['c']
uses[2] = ['a', []]
uses[3] = [[]] 
uses['test1'] = ['c', 'a', []] 

i dopiero teraz DFS wchodzi do procedury test2 i za każdym razem kiedy będzie dodawana jakaś zmienna do uses['test2'] dodawana też będzie do wszystkich tych pustych list. Jeśli będziemy chcieli od razu robić płaskie listy to zmienne z uses['test2'] nie będą się automatycznie dodawać

Reply:
> ale ten jezyk nie wspiera rekurencji.
A to myślałem że ma wspierać bo w main na masterze jest drzewo z rekurencją i Build .exe for Windows nie przechodziło

Reply:
> A to myślałem że ma wspierać bo w main na masterze jest drzewo z rekurencją i Build .exe for Windows nie przechodziło

Czyli możemy w sumie pominac te fory, tak? No bo technicznie patrzac to te ify jedynie pomijaja rekurencje. Jak je usuniemy to zostaje:

```suggestion
    for key, value in modifies.items():
        modifies[key] = list(set(value))

    for key, value in uses.items():
        uses[key] = list(set(value))
```

Co wydaje sie zbedne na pierwszy rzut oka. Samo `uses` i `modifies` mogloby zamiast list miec od poczatku sety. Wtedy po dodaniu tego samego elementu i tak bylby tylko jeden w srodku. Nie musielibysmy iterowac po nich i sprzatac.

Reply:
Nie są zbędne ponieważ możemy nie mieć rekurencji ale jeśli procedura test1 wywołuje test2 to słownik modifes bedzie przykładowo wyglądać tak: modifies["test1"] = ["a", "b", "c", [ _zmienne z procedury test2_ ]] więc i tak trzeba przelecieć po wartościach każdego klucza aby sprawdzić czy nie ma tam tablic

Reply:
To zamiast trzymac zmienne procedury `test2`, nie mozemy po prostu trzymac node procedury `test2`? Potem podczas lookupu, jak trafimy na ten node, to bedziemy wchodzic do niego. Takie podejscie bedzie prostsze w logice, zajmowalo mniej pamieci, a zlozonosciowo bedzie lzejsze o te fory.

Reply:
Raczej nie bo jak bedziemy mieli zapytanie:
`while w1 ; Select w1 such that Modifies(w1, "a")`
będziemy musieli przeglądać w słowniku wartości dla wszystkich statementów czy nie ma tam node procedure, i dla wszystkich znalezionych procedur jeszcze przeglądać wartości dla tej procedury czy nie ma tam zmiennej "a"

Reply:
W najnowszej zmianie mamy tak że przechodzimy po słowniku i dopisujemy zmienne z procedur w miejsca które trzeba bez przeglądania wartości słownika, a następnie podczas przetwarzania zapytanie sprawdzamy czy w wartościach jakiegoś klucza nie ma zmiennej "a". A tak przetwarzamy zapytanie sprawdzając czy w wartościach jakiegoś klucza nie ma zmiennej "a", a jak natrafimy na procedure node czy w tej procedurze nie ma zmiennej "a".
W pierwszym przypadku dopisujemy wartości z kluczy procedur x razy, a w drugim przypadku przeglądamy dodatkowo wartości procedur x razy, więc tak naprawdę wychodzi na to samo tylko zależy co jest bardziej optymalne dopisywanie czy sprawdzanie.

Reply:
I chyba w wymaganiach projektu jest żeby stworzyć słownik z uwzględnieniem call, a nie wiem czy to się liczy jako stworzenie słownika jeśli w wartościach będę inne elementy niż zmienne